### ISSUE TITLE
[Phase 2 Parity] AdamW, Learning Rate Schedulers, and Scheduler Architecture Refactor

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#388]
## Overview

This issue covers AdamW implementation, learning rate scheduler integration into ALL gradient-based optimizers, and a broader scheduler architecture refactor to properly organize diffusion and curriculum schedulers.

### Implementation
- Create `AdamWOptimizer<T>` as a **separate class** inheriting from `GradientBasedOptimizerBase<T>`
- Add `weightDecay` as a **separate parameter** (NOT L2 regularization - AdamW uses decoupled weight decay)
- Follow existing gradient-based optimizer pattern in `src/Optimizers/`

### Research-Based Defaults (PyTorch/HuggingFace conventions)
- `learningRate`: 1e-3
- `beta1`: 0.9
- `beta2`: 0.999
- `epsilon`: 1e-8
- `weightDecay`: 0.01 (industry standard, though PyTorch defaults to 0)

### Architecture Decision
**LR schedulers will be nullable parameters in optimizer constructors** (not a separate facade method).

### Base Class Changes (`GradientBasedOptimizerBase<T>`)

Add to the base class:
```csharp
// New fields
protected ILearningRateScheduler? _learningRateScheduler;
protected SchedulerStepMode _schedulerStepMode;
protected int _currentStep = 0;
protected int _currentEpoch = 0;

// New methods
public void StepScheduler()  // Called based on step mode
public void OnEpochEnd()     // Called at end of each epoch
public void OnBatchEnd()     // Called at end of each batch
```

### Constructor Pattern for ALL Optimizers
Each gradient-based optimizer gets scheduler parameters:
```csharp
public AdamOptimizer(
    IFullModel<T, TInput, TOutput>? model,
    AdamOptimizerOptions<T, TInput, TOutput> options,
    ILearningRateScheduler? scheduler = null,  // NEW
    SchedulerStepMode stepMode = SchedulerStepMode.StepPerEpoch)  // NEW
```

### SchedulerStepMode Enum
```csharp
public enum SchedulerStepMode
{
    StepPerBatch,      // Call Step() after each mini-batch
    StepPerEpoch,      // Call Step() after each epoch (default)
    WarmupThenEpoch    // Warmup steps per-batch, then per-epoch
}
```
> Note: Manual mode excluded to keep facade clean

### Complete List of Gradient-Based Optimizers to Update (25 total)

| # | Optimizer | Default Scheduler | Rationale |
|---|-----------|------------------|-----------|
| 1 | AdamOptimizer | None (constant LR) | Adaptive, rarely needs scheduler |
| 2 | AdamWOptimizer (NEW) | LinearWarmupScheduler | HuggingFace convention for transformers |
| 3 | AdagradOptimizer | None | Self-adapting learning rates |
| 4 | AdaDeltaOptimizer | None | Adaptive, no LR needed |
| 5 | AdaMaxOptimizer | None | Adaptive variant of Adam |
| 6 | AMSGradOptimizer | None | Adaptive like Adam |
| 7 | NadamOptimizer | None | Nesterov + Adam, adaptive |
| 8 | LionOptimizer | CosineAnnealingLR | Google's Lion benefits from scheduling |
| 9 | StochasticGradientDescentOptimizer | StepLRScheduler | Classic choice for SGD |
| 10 | GradientDescentOptimizer | StepLRScheduler | Benefits from decay |
| 11 | MiniBatchGradientDescentOptimizer | StepLRScheduler | Benefits from decay |
| 12 | MomentumOptimizer | StepLRScheduler | Benefits from decay |
| 13 | NesterovAcceleratedGradientOptimizer | StepLRScheduler | Benefits from decay |
| 14 | RootMeanSquarePropagationOptimizer | None | Adaptive |
| 15 | ConjugateGradientOptimizer | None | Line search handles step size |
| 16 | BFGSOptimizer | None | Quasi-Newton, self-adjusting |
| 17 | LBFGSOptimizer | None | Limited-memory BFGS |
| 18 | DFPOptimizer | None | Quasi-Newton method |
| 19 | NewtonMethodOptimizer | None | Uses Hessian for step size |
| 20 | LevenbergMarquardtOptimizer | None | Trust-region adjusts step |
| 21 | TrustRegionOptimizer | None | Self-adjusting trust region |
| 22 | CoordinateDescentOptimizer | None | Coordinate-wise updates |
| 23 | ProximalGradientDescentOptimizer | StepLRScheduler | Benefits from decay |
| 24 | ADMMOptimizer | None | Alternating direction |
| 25 | FTRLOptimizer | None | Follow-the-regularized-leader |

### Wiring Details

1. **Base class modification**: Add `ILearningRateScheduler?` field and step logic to `GradientBasedOptimizerBase`
2. **Options class update**: Add scheduler and step mode to `GradientBasedOptimizerOptions`
3. **Scheduler step integration**:
   - In `UpdateParameters()` methods, after updating LR call `_learningRateScheduler?.Step(_currentStep, _currentLearningRate)`
   - Track `_currentStep` and `_currentEpoch` for scheduler state
4. **LR update**: Before each parameter update, get LR from scheduler: `_currentLearningRate = _learningRateScheduler?.GetLearningRate(_currentStep) ?? _currentLearningRate`

### Existing LR Schedulers (already exist in `src/LearningRateSchedulers/`)
- [x] StepLRScheduler
- [x] CosineAnnealingLRScheduler
- [x] ExponentialLRScheduler
- [x] ReduceOnPlateauScheduler
- [x] CyclicLRScheduler
- [x] OneCycleLRScheduler
- [x] LinearWarmupScheduler
- [x] LearningRateSchedulerFactory

### Move and Rename
- **Move**: `src/Diffusion/` -> `src/NeuralNetworks/Diffusion/`
- **Update namespace**: `AiDotNet.Diffusion` -> `AiDotNet.NeuralNetworks.Diffusion`
- **Rename interface**: `IStepScheduler<T>` -> `INoiseScheduler<T>` (avoids confusion with LR schedulers)

### Affected Files
- `src/Diffusion/Schedulers/IStepScheduler.cs` -> `INoiseScheduler.cs`
- `src/Diffusion/Schedulers/DDIMScheduler.cs`
- `src/Diffusion/Schedulers/PNDMScheduler.cs`
- All diffusion model files referencing the old namespace

### Same Pattern as LR Schedulers
Add `ICurriculumScheduler?` as nullable parameter to relevant training components (e.g., `CurriculumLearner`).

### Existing Curriculum Schedulers (in `src/CurriculumLearning/Schedulers/`)
- LinearScheduler
- SelfPacedScheduler
- CompetenceBasedScheduler

### Phase A: AdamW (Priority 1)
- [ ] Create `AdamWOptimizer<T, TInput, TOutput>` with decoupled weight decay
- [ ] Create `AdamWOptimizerOptions<T, TInput, TOutput>`
- [ ] Add unit tests for AdamW
- [ ] Update optimizer factory

### Phase B: Base Class LR Scheduler Support (Priority 2)
- [ ] Add `SchedulerStepMode` enum to `src/Enums/`
- [ ] Add `ILearningRateScheduler?` field to `GradientBasedOptimizerBase`
- [ ] Add `SchedulerStepMode` field to `GradientBasedOptimizerBase`
- [ ] Add step tracking fields (`_currentStep`, `_currentEpoch`)
- [ ] Implement `StepScheduler()`, `OnEpochEnd()`, `OnBatchEnd()` methods
- [ ] Update `GradientBasedOptimizerOptions` to include scheduler params
- [ ] Integrate LR update into `UpdateParameters()` methods

### Phase C: Update All 25 Optimizers (Priority 3)
- [ ] AdamOptimizer - add scheduler params
- [ ] AdagradOptimizer - add scheduler params
- [ ] AdaDeltaOptimizer - add scheduler params
- [ ] AdaMaxOptimizer - add scheduler params
- [ ] AMSGradOptimizer - add scheduler params
- [ ] NadamOptimizer - add scheduler params
- [ ] LionOptimizer - add scheduler params
- [ ] StochasticGradientDescentOptimizer - add scheduler params
- [ ] GradientDescentOptimizer - add scheduler params
- [ ] MiniBatchGradientDescentOptimizer - add scheduler params
- [ ] MomentumOptimizer - add scheduler params
- [ ] NesterovAcceleratedGradientOptimizer - add scheduler params
- [ ] RootMeanSquarePropagationOptimizer - add scheduler params
- [ ] ConjugateGradientOptimizer - add scheduler params
- [ ] BFGSOptimizer - add scheduler params
- [ ] LBFGSOptimizer - add scheduler params
- [ ] DFPOptimizer - add scheduler params
- [ ] NewtonMethodOptimizer - add scheduler params
- [ ] LevenbergMarquardtOptimizer - add scheduler params
- [ ] TrustRegionOptimizer - add scheduler params
- [ ] CoordinateDescentOptimizer - add scheduler params
- [ ] ProximalGradientDescentOptimizer - add scheduler params
- [ ] ADMMOptimizer - add scheduler params
- [ ] FTRLOptimizer - add scheduler params
- [ ] AdamWOptimizer (NEW) - add scheduler params

### Phase D: Diffusion Refactor (Priority 4)
- [ ] Move `src/Diffusion/` to `src/NeuralNetworks/Diffusion/`
- [ ] Update namespace to `AiDotNet.NeuralNetworks.Diffusion`
- [ ] Rename `IStepScheduler<T>` to `INoiseScheduler<T>`
- [ ] Update all references across codebase
- [ ] Update any using statements

### Phase E: Curriculum Integration (Priority 5)
- [ ] Add `ICurriculumScheduler?` to `CurriculumLearner` constructor
- [ ] Ensure consistent pattern with LR schedulers
- [ ] Update relevant training loops

### Phase F: Testing
- [ ] Unit tests for all scheduler integrations
- [ ] Integration tests for optimizer + scheduler combinations
- [ ] Verify backwards compatibility (null scheduler = current behavior)

## References
- [PyTorch Optimizers](https://docs.pytorch.org/docs/stable/optim.html)
- [PyTorch LR Schedulers](https://docs.pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate)
- [HuggingFace Optimization](https://huggingface.co/docs/transformers/main_classes/optimizer_schedules)
- [Super-Convergence Paper](https://arxiv.org/abs/1708.07120) (OneCycleLR)
- [AdamW Paper](https://arxiv.org/abs/1711.05101) (Decoupled Weight Decay)
